[Feat] 1인 학습 (#64), index, study-solo 페이지 by Thymeleaf + tailwind.css (#66), Problem, Reply의 Request, Response -> Record로 교체 (#68)

sseunghoon opened the pull request:
## 주요 변경 사항
- 1인 학습 기능 구현 #64 
- index, study-solo 페이지 by Thymeleaf + tailwind.css #66 
- Reply의 Request, Response -> Record로 교체 #68 
- Reply, History 쿼리 최적화 (FetchJoin), Reply (un)recommend 로직 변경 #73

sseunghoon committed:
feat: home, study-solo 페이지 Thymeleaf + tailwind.css (레이아웃 구성)

defaultLayout 속 config.html, header.html, footer.html 등

signUp.html 과
login.html 은 테스트를 위한 페이지로 서비스에선 제거하거나 어드민 페이지로 옮겨야 함

sseunghoon committed:
feat: Keyword -> Problem 일방향 연관관계 에서 양방향 연관관계로 수정 및 Problem+Keyword FetchJoin

sseunghoon committed:
refactor: Problem Request, Response Record로 정리

sseunghoon committed:
refac: Reply Request, Response Record로 정리 및 필요없는 의존성, 함수 삭제

sseunghoon committed:
refac: Unused import, dependency, function 삭제

sseunghoon committed:
test: ProblemRegisterRequestDto -> RegisterProblemRequest 변경 및 테스트 보완

its-sky commented on src/main/java/cos/peerna/domain/history/controller/HistoryController.java:
```diff
@@ -20,9 +19,4 @@ public class HistoryController {
         return ResponseEntity.ok(historyService.findUserHistory(user, page));
     }
-
-    @GetMapping("/detail")
```
이건 왜 지우신건가요?

sseunghoon replied:
감사해요 History에 대한 개편을 생각하면서 지웠다가
생각을 바꾸면서 살려놨어야 하는 함수였습니다. 꼼꼼히 봐주셔서 감사해요~

its-sky commented on src/main/java/cos/peerna/domain/problem/dto/response/GetAnswerAndKeywordResponse.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.validation.annotation.Validated;
+
+@Validated
+public record GetAnswerAndKeywordResponse(
```
이거 네이밍이 앞에 Get이 붙을 필요가 없어보이긴 합니다

sseunghoon replied:
네 반영하겠습니다~

its-sky commented on src/main/java/cos/peerna/domain/problem/dto/response/GetProblemResponse.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.validation.annotation.Validated;
+
+@Validated
+public record GetProblemResponse(
```
요것도 마찬가지로 생각해보시면 좋을거 같아요

its-sky commented on src/main/java/cos/peerna/domain/problem/model/ProblemAnswerKeywords.java:
```diff
@@ -0,0 +1,9 @@
+public interface ProblemAnswerKeywords {
+    String getAnswer();
+    List<Keyword> getKeywords();
+}
```
이건 어디에 쓰이는 인터페이스인가요?

sseunghoon replied:
JPA Projection 할 때 쓰던건데 뭔가 이상한가요?

its-sky commented on src/main/java/cos/peerna/domain/problem/repository/ProblemRepository.java:
```diff
@@ -12,6 +14,7 @@
-    Long countByCategory(Category category);
-    List<ProblemIdMapping> findAllByCategory(Category category);
+    @Query("SELECT p.answer AS answer, k AS keywords FROM Problem p LEFT JOIN p.keywords k WHERE p.id = :id")
+    Optional<ProblemAnswerKeywords> findAnswerAndKeywordsById(Long id);
```
혹시 인터페이스가 JPA Projection을 할 때 사용하는 인터페이스인가요?

sseunghoon replied:
맞아요~ 이상한 부분이 있을까요?

its-sky commented on src/main/java/cos/peerna/domain/reply/model/Reply.java:
```diff
@@ -45,17 +46,27 @@ public class Reply {
 
     private Long likeCount;
```
혹시 이건 likes onetomany로 걸어두고 살려두신 이유가 있을까요?

sseunghoon replied:
데이터 정합성이 깨질 위험이 있을수도 있지만 여기서 정합성이 크리티컬한 부분이 아니라고 생각해서
조회 성능을 좋게하려고 한 의도였던거 같아요. 지우는게 좋을까요?

근데 FetchType.Lazy 가 추가가 안 되어있네요 likeCount를 지우든 안 지우든 일단 추가해야겠어요

its-sky commented:
그리고 한 PR에 수정 내용이 너무 커서 보기가 힘드네요..

sseunghoon commented:
> 그리고 한 PR에 수정 내용이 너무 커서 보기가 힘드네요..

피드백 감사해요 반영하겠습니다!

sseunghoon committed:
feat: Reply, History 쿼리 최적화 (FetchJoin), Reply (un)recommend 로직 변경

findByIdWithProblem, findByIdWithUserAndLike, findRepliesByUserOrderByIdDesc, findRepliesByProblemOrderByLikeCountDesc, findRepliesWithUserByHistoryOrderByHistoryIdDesc
